wmfdata/spark.py

```diff
@@ -112,20 +121,35 @@ def get_custom_session(
     """
     check_kerberos_auth()
 
+    if SPARK_HOME == "/usr/lib/spark2":
+        warnings.warn(
+            "\n    Spark 2 has been deprecated. Please migrate to Spark 3."
+            "\n    See https://wikitech.wikimedia.org/wiki/Analytics/Systems/Cluster/Spark/"
+            "Migration_to_Spark_3",
+            category=FutureWarning
+        )
+
+    old_session = get_active_session()
+    if old_session:
+        old_session.stop()
+
     if master == "yarn":
         if ship_python_env:
             # The path to our packed conda environment.
             conda_packed_file = conda.pack(**conda_pack_kwargs)
             # This will be used as the unpacked directory name in the YARN working directory.
             conda_packed_name = os.path.splitext(os.path.basename(conda_packed_file))[0]
 
-            # Ship conda_packed_file to each YARN worker.
+            # Ship conda_packed_file to each YARN worker, unless a previous session has done it already.
             conda_spark_archive = f"{conda_packed_file}#{conda_packed_name}"
-            if "spark.yarn.dist.archives" in spark_config:
+            if not old_session and "spark.yarn.dist.archives" in spark_config:
                 spark_config["spark.yarn.dist.archives"] += f",{conda_spark_archive}"
-            else:
+            elif not old_session:
```

Comment:
(If we do need to check `old_session`), you are currently are checking `not old_session` twice here. I'd find this more readable like:

```python
if not old_session:
    if "spark.yarn.dist.archives" in spark_config:
        # ...
    else:
        # ....
```

Reply:
Agreed, will fix.

Reply:
Fixed.